### PR TITLE
Create default config if none exists

### PR DESCRIPTION
**Pull request by nv-braf**
If no model config file is provided, MA will have Trition create one, if possible.
Moved server handler functions from entrypoint into there own dedicated class.

We need an L0 test to cover this, but I have observed this working correctly for resnet_v1_50_cpu_savedmodel (and taking an exception for resnet_libtorch)

### CODE REVIEW

File: model_analyzer/triton/server/server_handler.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+class TritonServerHandler():
+    """
+    Static methods used to create and return Triton Servers
+    """
+
+    def __init__(self):
+        NotImplemented
+
+    @staticmethod
+    def get_server_handle(config, gpus, use_model_repository=False):
```
Comment (tgerdesnv) [inline model_analyzer/triton/server/server_handler.py:39]: refactor this function

File: model_analyzer/triton/server/server_handler.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        if not path or not os.path.exists(path) or not os.path.isdir(path):
+            raise TritonModelAnalyzerException(
+                f"triton_install_path {path} is not specified, does not exist, " \
+                "or is not a directory."
+            )
+
+        # Make sure that files exist in the install directory
+        if len(os.listdir(path)) == 0:
+            raise TritonModelAnalyzerException(
+                f"triton_install_path {path} should not be empty.")
```
Comment (tgerdesnv) [inline model_analyzer/triton/server/server_handler.py:199]: endl

File: experiments/generator_experiment_factory.py [hunk 3]
```diff
@@ -39,7 +39,7 @@ def create_generator(generator_name, config_command):
         """
 
         if generator_name == "RunConfigGenerator":
-            generator = RunConfigGenerator(config_command,
+            generator = RunConfigGenerator(config_command, MagicMock(),
```
Comment (tgerdesnv) [inline experiments/generator_experiment_factory.py:42]: line 69 below will need to be updated. I pushed up the fix myself.

File: model_analyzer/config/generate/base_model_config_generator.py [hunk 4]
```diff
@@ -181,10 +197,63 @@ def make_model_config(param_combo, model, model_repository,
 
         return model_config
 
-    @staticmethod
-    def get_base_model_config_dict(model_repository, model_name):
-        config = ModelConfig.create_from_file(
-            f'{model_repository}/{model_name}')
+    @classmethod
+    def get_base_model_config_dict(cls, config, client, gpus, model_repository,
```
Comment (tgerdesnv) [inline model_analyzer/config/generate/base_model_config_generator.py:201]: I don't think this class should be the one who does all of this.  Can we push it down into ModelConfig class?

File: model_analyzer/config/generate/base_model_config_generator.py [hunk 5]
```diff
@@ -16,21 +16,29 @@
 
 from model_analyzer.constants import LOGGER_NAME
 from model_analyzer.triton.model.model_config import ModelConfig
+from model_analyzer.triton.server.server_handler import TritonServerHandler
+from model_analyzer.model_analyzer_exceptions import TritonModelAnalyzerException
+
+import os
 import abc
 import logging
+import requests
```
Comment (tgerdesnv) [inline model_analyzer/config/generate/base_model_config_generator.py:25]: not used

File: model_analyzer/config/generate/base_model_config_generator.py [hunk 6]
```diff
@@ -16,21 +16,29 @@
 
 from model_analyzer.constants import LOGGER_NAME
 from model_analyzer.triton.model.model_config import ModelConfig
+from model_analyzer.triton.server.server_handler import TritonServerHandler
+from model_analyzer.model_analyzer_exceptions import TritonModelAnalyzerException
+
+import os
```
Comment (tgerdesnv) [inline model_analyzer/config/generate/base_model_config_generator.py:22]: I wouldn't expect this file to need this

File: model_analyzer/config/generate/automatic_model_config_generator.py [hunk 7]
```diff
@@ -24,12 +24,13 @@
 class AutomaticModelConfigGenerator(BaseModelConfigGenerator):
     """ Given a model, generates model configs in automatic search mode """
 
-    def __init__(self, config, model, client, variant_name_manager,
+    def __init__(self, config, gpus, model, client, variant_name_manager,
```
Comment (tgerdesnv) [inline model_analyzer/config/generate/automatic_model_config_generator.py:27]: I don't like to see our constructor arg list continue to grow. If it is a necessary change, then we may want to refactor. I know on c++ land we would create a struct for stuff like this. Maybe the common info goes into a ConfigOptions struct that contains most of this? Or we piggy-back off of the SearchConfig class I recently created?

My main question is, do we actually need the GPUs passed around everywhere?

File: model_analyzer/triton/server/server_handler.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+from urllib.parse import urlparse
+
+import os
+import logging
+
+logger = logging.getLogger(LOGGER_NAME)
+
+
+class TritonServerHandler():
```
Comment (tgerdesnv) [inline model_analyzer/triton/server/server_handler.py:30]: class name should match file name.

File: model_analyzer/triton/server/server_handler.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+            Handle to the Triton Server
+        """
+
+        if config.triton_launch_mode == 'remote':
+            triton_config = TritonServerConfig()
+            triton_config.update_config(config.triton_server_flags)
+            triton_config['model-repository'] = 'remote-model-repository'
+            logger.info('Using remote Triton Server')
+            server = TritonServerFactory.create_server_local(
+                path=None, config=triton_config, gpus=[], log_path="")
```
Comment (tgerdesnv) [inline model_analyzer/triton/server/server_handler.py:65]: I think this line shows us that the GPUs is optional. As such, you could make the default "[ ]" and not have to plumb gpus everywhere.  That should allow you drop the number of files changed significantly.

File: model_analyzer/entrypoint.py [hunk 10]
```diff
@@ -17,22 +17,17 @@
...
-from .triton.server.server_factory import TritonServerFactory
-from .triton.server.server_config import TritonServerConfig
+from .triton.server.server_handler import TritonServerHandler
 from .triton.client.client_factory import TritonClientFactory
 from .state.analyzer_state_manager import AnalyzerStateManager
 from .config.input.config_command_profile import ConfigCommandProfile
 from .config.input.config_command_analyze import ConfigCommandAnalyze
 from .config.input.config_command_report import ConfigCommandReport
 from .log_formatter import setup_logging
-from model_analyzer.config.input.config_utils import \
```
Comment (tgerdesnv) [inline model_analyzer/entrypoint.py:29]: I love seeing so much deleted code!! None of this should have ever been here, and I'm glad it is being moved to a more appropriate place!

File: model_analyzer/triton/server/server_handler.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+import os
+import logging
+
+logger = logging.getLogger(LOGGER_NAME)
+
+
+class TritonServerHandler():
+    """
+    Static methods used to create and return Triton Servers
```
Comment (tgerdesnv) [inline model_analyzer/triton/server/server_handler.py:32]: Is there much of a difference between the factory and the functionality of this class? Just thinking out loud and wondering if they should be combined.

**Review by tgerdesnv — CHANGES_REQUESTED** [review]

File: model_analyzer/entrypoint.py (hunk 10 above)
Comment (nv-braf) [reply model_analyzer/entrypoint.py:29]: Thanks!

File: model_analyzer/config/generate/base_model_config_generator.py (hunk 5 above)
Comment (nv-braf) [reply model_analyzer/config/generate/base_model_config_generator.py:25]: Removed

File: model_analyzer/config/generate/automatic_model_config_generator.py (hunk 7 above)
Comment (tgerdesnv) [reply model_analyzer/config/generate/automatic_model_config_generator.py:27]: I still don't think that any of the generators will ever need GPUs. We can create the temporary triton server without GPUs.

File: model_analyzer/triton/server/server_handler.py (hunk 1 above)
Comment (nv-braf) [reply model_analyzer/triton/server/server_handler.py:39]: Done

File: model_analyzer/triton/server/server_handler.py (hunk 11 above)
Comment (nv-braf) [reply model_analyzer/triton/server/server_handler.py]: I've removed the handler and moved all the code into the factory

File: model_analyzer/config/generate/automatic_model_config_generator.py (hunk 7 above)
Comment (nv-braf) [reply model_analyzer/config/generate/automatic_model_config_generator.py:27]: We can, but it doesn't result in the same config being created - at a minimum we'd need to hack in and change from CPU to GPU_ONLY. Not sure if there are any other side effects, this seems unnecessarily risky.

File: model_analyzer/config/generate/base_model_config_generator.py (hunk 4 above)
Comment (nv-braf) [reply model_analyzer/config/generate/base_model_config_generator.py:196]: Done

File: model_analyzer/config/generate/base_model_config_generator.py [hunk 12]
```diff
@@ -16,6 +16,8 @@
 
 from model_analyzer.constants import LOGGER_NAME
 from model_analyzer.triton.model.model_config import ModelConfig
+from model_analyzer.model_analyzer_exceptions import TritonModelAnalyzerException
```
Comment (tgerdesnv) [inline model_analyzer/config/generate/base_model_config_generator.py:19]: is this still needed?

File: model_analyzer/config/generate/base_model_config_generator.py [hunk 13]
```diff
@@ -181,11 +192,28 @@ def make_model_config(param_combo, model, model_repository,
...
-    @staticmethod
-    def get_base_model_config_dict(model_repository, model_name):
-        config = ModelConfig.create_from_file(
-            f'{model_repository}/{model_name}')
-        return config.get_config()
+    @classmethod
+    def get_base_model_config_dict(cls, config, client, gpus, model_repository,
+                                   model_name):
+        """ 
+        Attempts to create a base model config dict from config.pbtxt, if one exists
```
Comment (tgerdesnv) [inline model_analyzer/config/generate/base_model_config_generator.py:199]: This comment should go to wherever the actual code went.

File: model_analyzer/triton/model/model_config.py [hunk 14]
```diff
@@ -56,6 +60,84 @@ def from_dict(cls, model_config_dict):
 
         return model_config
 
+    @staticmethod
+    def create_model_config_dict(config, client, gpus, model_repository,
```
Comment (tgerdesnv) [inline model_analyzer/triton/model/model_config.py:64]: comment these two functions

**Review by tgerdesnv — CHANGES_REQUESTED** [review]

File: model_analyzer/config/generate/base_model_config_generator.py (hunk 12 above)
Comment (nv-braf) [reply model_analyzer/config/generate/base_model_config_generator.py:19]: Removed

File: model_analyzer/triton/model/model_config.py (hunk 14 above)
Comment (nv-braf) [reply model_analyzer/triton/model/model_config.py:64]: Done

File: model_analyzer/config/generate/base_model_config_generator.py (hunk 13 above)
Comment (nv-braf) [reply model_analyzer/config/generate/base_model_config_generator.py:198]: Agreed. I've left it here and also replicated it for create_model_config_dict

File: model_analyzer/config/generate/base_model_config_generator.py [hunk 15]
```diff
@@ -181,10 +197,63 @@ def make_model_config(param_combo, model, model_repository,
...
+
+            client.wait_for_model_ready(model_name, config.client_max_retries)
+
+            cls._default_config[model_name] = client.get_model_config(
+                model_name, config.client_max_retries)
+            server.stop()
+
+            return cls._default_config[model_name]
+
         return config.get_config()
```
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply model_analyzer/config/generate/base_model_config_generator.py]: I created a ticket

**Review by tgerdesnv — APPROVED** [review]

**Pull request merged**